judge/views/submission.py

```diff
@@ -235,10 +242,15 @@ def access_check(self, request):
             if not user.is_authenticated():
                 raise Http404()
 
-            if (not self.problem.authors.filter(id=user.profile.id).exists() and
-                    not user.has_perm('judge.see_private_problem') and
-                    not (self.in_contest and self.contest.problems.filter(id=self.problem.id).exists())):
-                raise Http404()
+            if self.problem.is_editor(user.profile) or self.problem.testers.filter(id=user.profile.id).exists():
+                return
+
+            if user.has_perm('judge.see_private_problem'):
+                return
+
```

Comment:
Yeah, it kinda looked worse before. Not that I have any strong feelings on the matter.